hummingbot/connector/exchange/xrpl/xrpl_api_user_stream_data_source.py

```diff
@@ -0,0 +1,70 @@
+import asyncio
+import time
+from typing import TYPE_CHECKING, Any, Dict, Optional
+
+from xrpl.asyncio.clients import AsyncWebsocketClient
+from xrpl.models import Subscribe
+from xrpl.models.requests import StreamParameter
+
+from hummingbot.connector.exchange.xrpl.xrpl_auth import XRPLAuth
+from hummingbot.core.data_type.user_stream_tracker_data_source import UserStreamTrackerDataSource
+from hummingbot.logger import HummingbotLogger
+
+if TYPE_CHECKING:
+    from hummingbot.connector.exchange.xrpl.xrpl_exchange import XrplExchange
+
+
+class XRPLAPIUserStreamDataSource(UserStreamTrackerDataSource):
+    _logger: Optional[HummingbotLogger] = None
+
+    def __init__(self,
+                 auth: XRPLAuth,
+                 connector: 'XrplExchange'):
+        super().__init__()
+        self._connector = connector
+        self._auth = auth
+        self._xrpl_client = AsyncWebsocketClient(self._connector.node_url)
+        self._last_recv_time: float = 0
+
+    @property
+    def last_recv_time(self) -> float:
+        """
+        Returns the time of the last received message
+
+        :return: the timestamp of the last received message in seconds
+        """
+        return self._last_recv_time
+
+    async def listen_for_user_stream(self, output: asyncio.Queue):
+        """
+        Connects to the user private channel in the exchange using a websocket connection. With the established
+        connection listens to all balance events and order updates provided by the exchange, and stores them in the
+        output queue
+
+        :param output: the queue to use to store the received messages
+        """
+        while True:
+            try:
+                subscribe = Subscribe(accounts=[self._auth.get_account()], streams=[StreamParameter.TRANSACTIONS])
```

Comment:
One comment from me, what’s the reason for subscribing to `streams=[StreamParameter.TRANSACTIONS] `? This way you receive all transactions where I think only those specific to the account are important here.

Reply:
Because the subscription is for user stream which feeds event to `async def _user_stream_event_listener(self):` in `xrpl_exchange.py`. We only needs events that notify changes about user balance and order status which `StreamParameter.TRANSACTIONS` should have the information we need, for other kind of event, it is meaningless to include them.

Reply:
Right, but then `accounts=[self._auth.get_account()]` is sufficient as that would be for all transactions affecting that specific account (payments, offer changes etc.)

Reply:
You are right, I see it is not needed to include `treamParameter.TRANSACTIONS`, will remove it.